lib/pkce.js

```diff
@@ -0,0 +1,118 @@
+/*!
+ * Copyright (c) 2019-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ *
+ */
+
+ /* eslint-disable complexity, max-statements */
+var AuthSdkError  = require('./errors/AuthSdkError');
+var http          = require('./http');
+var util          = require('./util');
+
+// Code verifier: Random URL-safe string with a minimum length of 43 characters.
+// Code challenge: Base64 URL-encoded SHA-256 hash of the code verifier.
+var MIN_VERIFIER_LENGTH = 43;
+var MAX_VERIFIER_LENGTH = 128;
+
+function generateVerifier(prefix) {
+  var verifier = prefix || '';
+  if (verifier.length < MIN_VERIFIER_LENGTH) {
+    verifier = verifier + util.genRandomString(MIN_VERIFIER_LENGTH - verifier.length);
+  }
+  return encodeURIComponent(verifier).slice(0, MAX_VERIFIER_LENGTH);
+}
+
+function saveMeta(sdk, meta) {
+  var storage = sdk.options.storageUtil.getPKCEStorage();
+  storage.setStorage(meta);
+}
+
+function loadMeta(sdk) {
+  var storage = sdk.options.storageUtil.getPKCEStorage();
+  var obj = storage.getStorage();
+  return obj;
+}
+
+function clearMeta(sdk) {
+  var storage = sdk.options.storageUtil.getPKCEStorage();
+  storage.clearStorage();
+}
+
+/* global Uint8Array, TextEncoder */
+function computeChallenge(str) {  
+  var buffer = new TextEncoder().encode(str);
+  return crypto.subtle.digest('SHA-256', buffer).then(function(arrayBuffer) {
+    var hash = String.fromCharCode.apply(null, new Uint8Array(arrayBuffer));
+    var b64u = util.stringToBase64Url(hash); // url-safe base64 variant
+    return b64u;
+  });
+}
+
+
+function validateOptions(oauthOptions) {
+  // Quick validation
+  if (!oauthOptions.clientId) {
+    throw new AuthSdkError('A clientId must be specified in the OktaAuth constructor to get a token');
+  }
+
+  if (!oauthOptions.redirectUri) {
+    throw new AuthSdkError('The redirectUri passed to /authorize must also be passed to /token');
+  }
+
+  if (!oauthOptions.authorizationCode) {
+    throw new AuthSdkError('An authorization code (returned from /authorize) must be passed to /token');
+  }
+
+  if (!oauthOptions.codeVerifier) {
+    throw new AuthSdkError('The "codeVerifier" (generated and saved by your app) must be passed to /token');
+  }
+
+  if (oauthOptions.grantType !== 'authorization_code') {
+    throw new AuthSdkError('Expecting "grantType" to equal "authorization_code"');
```

Comment:
I could clone the oauthOptions and then set params on it. I decided to go with a simpler approach and use / validate the options without any cloning or overriding.

Reply:
Was this updated and not pushed to this topic branch?

Currently we're still cloning the `oauthParams` and extending them with default values before calling `pkce.getToken`, meaning most of these assertions can never fail.

Reply:
The way we are calling it from token, yes (I would hope) it would not fail. These specific assertions match server-side requirements for the /token endpoint specifically. (Whereas token.js concerns itself with the /authorize endpoint) If any one of these parameters is missing the POST will fail.

Reply:
If you'd prefer to handle the assertions here, please remove or modify the `clone` call before `pkce.getToken` is called. Checking this a second time isn't needed.